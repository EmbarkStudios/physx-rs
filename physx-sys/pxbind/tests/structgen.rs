fn gen_structgen(which: &str, _to_emit: &'static [&str]) -> anyhow::Result<String> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let record_filter = |rb: &pxbind::consumer::RecBinding<'_>| rb.name.starts_with("Px");

    let generator = pxbind::generator::Generator {
        record_filter: Box::new(record_filter),
        ..Default::default()
    };

    let structgen = {
        let mut sg = Vec::new();
        generator.generate_structgen(&consumer, &mut sg)?;
        String::from_utf8(sg)?
    };

    Ok(structgen)
}

struct StructgenOut {
    cpp: String,
    rust: String,
}

fn run_structgen(sg: String) -> anyhow::Result<StructgenOut> {
    let td = tempfile::tempdir()?;
    let tp = td.path();
    let sg_src = tp.join("structgen.cpp");

    std::fs::write(&sg_src, sg)?;
    std::fs::copy("tests/structgen.hpp", tp.join("structgen.hpp"))?;

    // Compile the structgen program for the host target
    let mut cmd = std::process::Command::new("clang++");
    cmd.current_dir(&td);

    let include_dir = pxbind::get_include_dir()?;
    cmd.args([
        "-fuse-ld=lld",
        "-std=c++11",
        "-fcolor-diagnostics",
        "-o",
        "structgen",
        "-DNDEBUG",
        "-I",
    ]);
    cmd.arg(include_dir);
    cmd.arg(&sg_src);

    anyhow::ensure!(cmd.status()?.success(), "failed to compile structgen");

    let mut sg = std::process::Command::new(tp.join("structgen"));
    sg.current_dir(tp);

    anyhow::ensure!(sg.status()?.success(), "failed to run structgen");

    let cpp = std::fs::read_to_string(tp.join("structgen_out.hpp"))?;
    let rust = std::fs::read_to_string(tp.join("structgen_out.rs"))?;

    Ok(StructgenOut { cpp, rust })
}

#[test]
fn many_things() {
    let sg = gen_structgen("functions/methods_complex.h", &[]).unwrap();

    insta::assert_snapshot!(sg);

    let generated = run_structgen(sg).unwrap();

    insta::assert_snapshot!(generated.cpp);
    insta::assert_snapshot!(generated.rust);
}

#[test]
fn all_the_things() {
    let sg = gen_structgen("all_the_things.h", &[]).unwrap();

    insta::assert_snapshot!(sg);

    let generated = run_structgen(sg).unwrap();

    insta::assert_snapshot!(generated.cpp);
    insta::assert_snapshot!(generated.rust);
}
