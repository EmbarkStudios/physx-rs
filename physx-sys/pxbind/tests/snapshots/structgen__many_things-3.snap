---
source: physx-sys/pxbind/tests/structgen.rs
expression: generated.rust
---
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxVec3 {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxVec4 {
    pub x: f32,
    pub y: f32,
    pub z: f32,
    pub w: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxQuat {
    pub x: f32,
    pub y: f32,
    pub z: f32,
    pub w: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxMat33 {
    pub column0: glam::Vec3,
    pub column1: glam::Vec3,
    pub column2: glam::Vec3,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxMat44 {
    pub column0: glam::Vec4,
    pub column1: glam::Vec4,
    pub column2: glam::Vec4,
    pub column3: glam::Vec4,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxTransform {
    pub q: glam::Quat,
    pub p: glam::Vec3,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxVec3Padded {
    pub x: f32,
    pub y: f32,
    pub z: f32,
    pub padding: u32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxTransformPadded {
    pub transform: PxTransform,
    pub padding: u32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxPlane {
    pub n: glam::Vec3,
    pub d: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxBoxGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub halfExtents: glam::Vec3,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxSphereGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub radius: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxCapsuleGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub radius: f32,
    pub halfHeight: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxPlaneGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxMeshScale {
    pub scale: glam::Vec3,
    pub rotation: PxQuat,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxStridedData {
    pub stride: u32,
    pub structgen_pad0: [u8; 4],
    pub data: *const std::ffi::c_void,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxBoundedData {
    pub stride: u32,
    pub structgen_pad0: [u8; 4],
    pub data: *const std::ffi::c_void,
    pub count: u32,
    pub structgen_pad1: [u8; 4],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxHullPolygon {
    pub mPlane: [f32; 4],
    pub mNbVerts: u16,
    pub mIndexBase: u16,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxConvexMeshGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub scale: PxMeshScale,
    pub structgen_pad1: [u8; 4],
    pub convexMesh: *mut PxConvexMesh,
    pub meshFlags: PxConvexMeshGeometryFlags,
    pub structgen_pad2: [u8; 7],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxTriangleMeshGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub scale: PxMeshScale,
    pub meshFlags: PxMeshGeometryFlags,
    pub structgen_pad1: [u8; 3],
    pub triangleMesh: *mut PxTriangleMesh,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxHeightFieldGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub heightField: *mut PxHeightField,
    pub heightScale: f32,
    pub rowScale: f32,
    pub columnScale: f32,
    pub heightFieldFlags: PxMeshGeometryFlags,
    pub structgen_pad1: [u8; 3],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxBounds3 {
    pub minimum: glam::Vec3,
    pub maximum: glam::Vec3,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxParticleSystemGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub mSolverType: PxParticleSolverType,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxHairSystemGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxTetrahedronMeshGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub tetrahedronMesh: *mut PxTetrahedronMesh,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxQueryHit {
    pub faceIndex: u32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxLocationHit {
    pub faceIndex: u32,
    pub flags: PxHitFlags,
    pub structgen_pad0: [u8; 2],
    pub position: glam::Vec3,
    pub normal: glam::Vec3,
    pub distance: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxGeomRaycastHit {
    pub faceIndex: u32,
    pub flags: PxHitFlags,
    pub structgen_pad0: [u8; 2],
    pub position: glam::Vec3,
    pub normal: glam::Vec3,
    pub distance: f32,
    pub u: f32,
    pub v: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxGeomOverlapHit {
    pub faceIndex: u32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxGeomSweepHit {
    pub faceIndex: u32,
    pub flags: PxHitFlags,
    pub structgen_pad0: [u8; 2],
    pub position: glam::Vec3,
    pub normal: glam::Vec3,
    pub distance: f32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxGeomIndexPair {
    pub id0: u32,
    pub id1: u32,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxQueryThreadContext {
    pub structgen_pad0: [u8; 1],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxCustomGeometry {
    pub structgen_pad0: [u8; 4],
    pub mTypePadding: f32,
    pub callbacks: *mut PxCustomGeometry::Callbacks,
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxGeometryHolder {
    pub structgen_pad0: [u8; 56],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxAllocator {
    pub structgen_pad0: [u8; 1],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxRawAllocator {
    pub structgen_pad0: [u8; 1],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxVirtualAllocator {
    pub structgen_pad0: [u8; 16],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxUserAllocated {
    pub structgen_pad0: [u8; 1],
}
#[derive(Clone, Copy)]
#[repr(C)]
pub struct PxShape {
    pub structgen_pad0: [u8; 16],
    pub userData: *mut std::ffi::c_void,
}
#[cfg(test)]
mod sizes {
    #[test]
    fn check_sizes() {
        assert_eq!(std::mem::size_of::<PxVec3>(), 12);
        assert_eq!(std::mem::size_of::<PxVec4>(), 16);
        assert_eq!(std::mem::size_of::<PxQuat>(), 16);
        assert_eq!(std::mem::size_of::<PxMat33>(), 36);
        assert_eq!(std::mem::size_of::<PxMat44>(), 64);
        assert_eq!(std::mem::size_of::<PxTransform>(), 28);
        assert_eq!(std::mem::size_of::<PxGeometry>(), 8);
        assert_eq!(std::mem::size_of::<PxVec3Padded>(), 16);
        assert_eq!(std::mem::size_of::<PxTransformPadded>(), 32);
        assert_eq!(std::mem::size_of::<PxPlane>(), 16);
        assert_eq!(std::mem::size_of::<PxBoxGeometry>(), 20);
        assert_eq!(std::mem::size_of::<PxSphereGeometry>(), 12);
        assert_eq!(std::mem::size_of::<PxCapsuleGeometry>(), 16);
        assert_eq!(std::mem::size_of::<PxPlaneGeometry>(), 8);
        assert_eq!(std::mem::size_of::<PxMeshScale>(), 28);
        assert_eq!(std::mem::size_of::<PxStridedData>(), 16);
        assert_eq!(std::mem::size_of::<PxBoundedData>(), 24);
        assert_eq!(std::mem::size_of::<PxHullPolygon>(), 20);
        assert_eq!(std::mem::size_of::<PxConvexMeshGeometry>(), 56);
        assert_eq!(std::mem::size_of::<PxTriangleMeshGeometry>(), 48);
        assert_eq!(std::mem::size_of::<PxHeightFieldGeometry>(), 32);
        assert_eq!(std::mem::size_of::<PxBounds3>(), 24);
        assert_eq!(std::mem::size_of::<PxParticleSystemGeometry>(), 12);
        assert_eq!(std::mem::size_of::<PxHairSystemGeometry>(), 8);
        assert_eq!(std::mem::size_of::<PxTetrahedronMeshGeometry>(), 16);
        assert_eq!(std::mem::size_of::<PxQueryHit>(), 4);
        assert_eq!(std::mem::size_of::<PxLocationHit>(), 36);
        assert_eq!(std::mem::size_of::<PxGeomRaycastHit>(), 44);
        assert_eq!(std::mem::size_of::<PxGeomOverlapHit>(), 4);
        assert_eq!(std::mem::size_of::<PxGeomSweepHit>(), 36);
        assert_eq!(std::mem::size_of::<PxGeomIndexPair>(), 8);
        assert_eq!(std::mem::size_of::<PxQueryThreadContext>(), 1);
        assert_eq!(std::mem::size_of::<PxCustomGeometry>(), 16);
        assert_eq!(std::mem::size_of::<PxGeometryHolder>(), 56);
        assert_eq!(std::mem::size_of::<PxAllocator>(), 1);
        assert_eq!(std::mem::size_of::<PxRawAllocator>(), 1);
        assert_eq!(std::mem::size_of::<PxVirtualAllocator>(), 16);
        assert_eq!(std::mem::size_of::<PxUserAllocated>(), 1);
        assert_eq!(std::mem::size_of::<PxShape>(), 24);
    }
}

